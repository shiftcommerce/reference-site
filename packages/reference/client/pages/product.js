// Libraries
import { Component } from 'react'
import { connect } from 'react-redux'
import Head from 'next/head'

// Lib
import renderComponents from '../lib/render-components'
import { algoliaReduxWrapper } from '../lib/algolia-redux-wrapper'

// Components
import Loading from '../components/loading'
import ProductDisplay from '../components/products/display/product-display'

// Actions
import { addToCart } from '../actions/cart-actions'
import { readProduct } from '../actions/product-actions'

export class Product extends Component {
  constructor (props) {
    super(props)
    // TODO: Update to use redux
    this.state = {
      sku: '',
      variant: '',
      variantId: null,
      quantity: 1,
      price: '',
      stockAvailableLevel: 0
    }

    this.changeVariant = this.changeVariant.bind(this)
    this.addToBag = this.addToBag.bind(this)
  }

  static async getInitialProps ({ reduxStore, req, query }) {
    const { id } = query
    const isServer = !!req

    if (isServer) {
      await reduxStore.dispatch(readProduct(id))
    }

    return { id: id }
  }

  componentDidMount () {
    const { dispatch, id } = this.props
    dispatch(readProduct(id))
  }

  addToBag () {
    const { variantId, quantity } = this.state
    this.props.dispatch(addToCart(variantId, parseInt(quantity)))
  }

  changeVariant (e) {
    this.setState({
      stockAvailableLevel: parseInt(e.target.attributes['data-stock-available-level'].value, 10),
      sku: e.target.value,
      variant: e.target.text,
      variantId: e.target.attributes['data-variant-id'].value,
      price: e.target.attributes['data-price'].value
    })
  }

  render () {
    const { product, product: { loading, error, template, title } } = this.props

    if (loading) {
      return (
        <Loading />
      )
    } else if (error || Object.keys(product).length === 0) {
      return (
        <h1>Unable to load product.</h1>
      )
    } else {
      const templateSection = template && template.sections && template.sections.slice(-1).pop()
      const components = templateSection && templateSection.components
      const selectedVariant = product.variants.find(variant => variant.id === this.state.variantId)

      return (
        <>
          <Head>
            <title>{ title }</title>
          </Head>
          <ProductDisplay
            product={product}
            changeVariant={this.changeVariant}
            addToBag={this.addToBag}
            clickToBuy={this.clickToBuy}
            selectedVariant={selectedVariant}
          />
          { components && renderComponents(components) }
        </>
      )
    }
  }
}

function mapStateToProps (state) {
  const { menu, product, cart } = state
  return { menu, product, cart }
}

export default algoliaReduxWrapper(connect(mapStateToProps)(Product), Product)
