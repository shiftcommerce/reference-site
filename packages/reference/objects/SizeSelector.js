// Libraries
import { Component } from 'react'

class SizeSelector extends Component {
  constructor (props) {
    super(props)
    this.renderOptions = this.renderOptions.bind(this)
  }

  renderOptions (variants) {
    return (
      variants && variants.map((variant, idx) =>
        <option role='option' key={idx} value={variant.id} aria-setsize={variants.length} aria-posinset={idx + 1}>{ variant.size }</option>
      )
    )
  }

  render () {
    let {
      name,
      prompt,
      variants,
      ...otherProps
    } = this.props

    return (
      <select name={name} {...otherProps}>

        <option role='option' value='' aria-setsize={variants.length} aria-posinset='0' >{ prompt }</option>

        { this.renderOptions(variants) }

      </select>
    )
  }
}

export default SizeSelector
