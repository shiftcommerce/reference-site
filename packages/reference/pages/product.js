// Libraries
import { Component } from 'react'
import { connect } from 'react-redux'

// Components
import ProductDisplay from '../components/products/pdp/ProductDisplay'

// Actions
import { addToCart } from '../actions/cartActions'
import { readProduct } from '../actions/productActions'

export class Product extends Component {
  constructor (props) {
    super(props)
    this.state = {
      sku: '',
      size: '',
      quantity: 1,
      price: ''
    }

    this.changeQuantity = this.changeQuantity.bind(this)
    this.changeSize = this.changeSize.bind(this)
    this.addToBag = this.addToBag.bind(this)
  }

  static async getInitialProps ({ store, isServer, pathname, query }) {
    await store.dispatch(readProduct(query.id))
  }

  addToBag () {
    const { quantity, sku, size, price } = this.state
    const product = this.props.product

    if (quantity !== '' && sku !== '') {
      const lineItem = {
        sku: sku,
        title: product.title,
        size: size,
        quantity: parseInt(quantity),
        price: price,
        imageUrl: (product.asset_files[0] && product.asset_files[0].s3_url),
        productSku: product.sku,
        productID: product.id
      }
      this.props.dispatch(addToCart(lineItem))
    } else {
      /*eslint-disable */
      alert('Please select a size')
      /*eslint-enable */
    }
  }

  changeQuantity (e) {
    this.setState({ quantity: e.target.value })
  }

  changeSize (e) {
    this.setState({
      sku: e.target.value,
      size: e.target.options[e.target.selectedIndex].text,
      price: e.target.options[e.target.selectedIndex].dataset.price
    })
  }

  render () {
    const product = this.props.product
    const { loading, error } = this.props.product

    if (loading) {
      return (
        <h1>Loading product...</h1>
      )
    } else if (error || Object.keys(product).length === 0) {
      return (
        <h1>Unable to load product.</h1>
      )
    } else {
      return (
        <ProductDisplay product={product} changeQuantity={this.changeQuantity} changeSize={this.changeSize} addToBag={this.addToBag} {...this.state} />
      )
    }
  }
}

const mapStateToProps = (state) => {
  return {
    product: state.product.data
  }
}

export default connect(mapStateToProps)(Product)
