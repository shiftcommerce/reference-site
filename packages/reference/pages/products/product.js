// Libraries
import { Component } from 'react'
import withRedux from 'next-redux-wrapper'

// Components
import Layout from '../../components/Layout'
import ProductDisplay from '../../components/products/pdp/ProductDisplay'

// Objects
import Breadcrumb from '../../objects/Breadcrumb'

// Actions
import { addToCart } from '../../actions/cartActions'
import { readProduct } from '../../actions/productActions'
import { readCategories } from '../../actions/categoryActions'

// Utils
import { configureStore } from '../../utils/configureStore'

class Product extends Component {
  constructor () {
    super()
    this.state = {
      sku: '',
      size: '',
      quantity: 1,
      price: ''
    }

    this.changeQuantity = this.changeQuantity.bind(this)
    this.changeSize = this.changeSize.bind(this)
    this.addToBag = this.addToBag.bind(this)
  }

  static async getInitialProps ({ store, query }) {
    await store.dispatch(readCategories(store))
    if (query.id) {
      await store.dispatch(readProduct(query.id))
    } else {
      return { product: {} }
    }
  }

  addToBag () {
    let { quantity, sku, size, price } = this.state
    let { product } = this.props

    if (quantity !== '' && sku !== '') {
      const lineItem = {
        sku: sku,
        title: product.title,
        size: size,
        quantity: parseInt(quantity),
        price: price,
        imageUrl: (product.asset_files[0] && product.asset_files[0].url),
        productSku: product.sku,
        productID: product.id
      }
      this.props.dispatch(addToCart(lineItem))
    } else {
      /*eslint-disable */
      alert('Please select a size')
      /*eslint-enable */
    }
  }

  changeQuantity (e) {
    this.setState({ quantity: e.target.value })
  }

  changeSize (e) {
    this.setState({
      sku: e.target.value,
      size: e.target.options[e.target.selectedIndex].text,
      price: e.target.options[e.target.selectedIndex].dataset.price
    })
  }

  render () {
    let {
      product
    } = this.props
    let category = (product.categories && product.categories[0])

    const breadcrumbMenuTrail = [
      { id: category.id, title: category.title, canonical_path: `/categories/${category.id}`, page: '/categories/category' },
      { id: product.id, title: product.title, canonical_path: `/products/${product.id}`, page: '/products/product' }
    ]

    if (product.error || Object.keys(product).length === 0) {
      return (
        <Layout>
          <h1>Unable to load product.</h1>
        </Layout>
      )
    } else if (product.loading) {
      return (
        <Layout>
          <h1>Loading product...</h1>
        </Layout>
      )
    } else {
      return (
        <Layout>
          <div>
            <Breadcrumb trail={breadcrumbMenuTrail} />
          </div>

          <ProductDisplay product={product} changeQuantity={this.changeQuantity} changeSize={this.changeSize} addToBag={this.addToBag} {...this.state} />
        </Layout>
      )
    }
  }
}

const mapStateToProps = (state) => {
  return {
    cart: state.cart || {},
    product: state.product || {}
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    dispatch: dispatch
  }
}

export default withRedux(configureStore, mapStateToProps, mapDispatchToProps)(Product)
