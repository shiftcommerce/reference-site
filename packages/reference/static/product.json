{
  "id": 1,
  "title": "Textured Long T-Shirt",
  "description": "Combining comfort and practicality, this grey marl textured t-shirt is a go-to everyday essential. Featuring a v-neck and front pocket, this staple piece can be layered to create a range different looks.",
  "sku": "SKU2659078",
  "price": 99,
  "meta": {
    "colour": "Blue"
  },
  "variants": [
    {
      "id": 1,
      "sku": "123",
      "price": 10,
      "title": "Small",
      "size": "Size 8"
    },
    {
      "id": 2,
      "sku": "124",
      "price": 11,
      "title": "Medium",
      "size": "Size 10"
    },
    {
      "id": 3,
      "sku": "125",
      "price": 12,
      "title": "Large",
      "size": "Size 12"
    }
  ],
  "bundles": [
    {
      "id": 1,
      "title": "Mingo",
      "products": [
        {
          "id": 1,
          "title": "Stripe Shirt",
          "canonical_path": "/products/stripe-shirt",
          "asset_files": [
            {
              "id": 2,
              "url": "https://staging-matalanintegration-1452506760.s3.amazonaws.com/uploads/asset_file/asset_file/12917/S2623404_C146_Main.jpg"
            }
          ]
        },
        {
          "id": 2,
          "title": "Slip on Pump (no asset files example)",
          "canonical_path": "/products/slip-on-pump",
          "asset_files": []
        }
      ]
    }
  ],
  "asset_files": [
    {
      "id": 1,
      "url": "https://staging-matalanintegration-1452506760.s3.amazonaws.com/uploads/asset_file/asset_file/12671/S2623408_C29P_Alt2.jpg"
    }
  ]
}