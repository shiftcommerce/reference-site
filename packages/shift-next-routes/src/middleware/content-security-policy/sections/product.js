// libs
const { buildContentSecurityPolicy } = require('../build-content-security-policy')

/**
 * Enables the Product Page Content Security Policy
 * @param {Function} server - eg. express
 * @param {object} options - eg. imageHosts, styleHosts, scriptHosts, frameHosts, connectHosts
 */
const productPageContentSecurityPolicy = (server, options = {}) => {
  server.all(/(\/pages\/product.js)$/, (req, res, next) => {
    console.log('Product PAGE')
    // Build and set the CSP header on the response
    res.set('Content-Security-Policy', buildContentSecurityPolicy(options))
    // Call the next middleware in the stack
    next()
  })
}

module.exports = { productPageContentSecurityPolicy: productPageContentSecurityPolicy }
