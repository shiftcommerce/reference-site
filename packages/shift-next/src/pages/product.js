// Libraries
import React, { Component, Fragment } from 'react'
import Router from 'next/router'

// Lib
import renderComponents from '../lib/render-components'
import { suffixWithStoreName } from '../lib/suffix-with-store-name'

// Components
import { Loading } from '@shiftcommerce/shift-react-components/src/objects/loading'
import { ProductDisplay } from '@shiftcommerce/shift-react-components/src/components/products/display/product-display'

// Actions
import { addToCart, toggleMiniBag } from '../actions/cart-actions'
import { readProduct } from '../actions/product-actions'

// Config
import Config from '../lib/config'
import ApiClient from '../lib/api-client'

class ProductPage extends Component {
  constructor (props) {
    super(props)

    this.state = {
      variantId: null,
      quantity: 1
    }

    this.Head = Config.get().Head
    this.changeVariant = this.changeVariant.bind(this)
    this.addToBag = this.addToBag.bind(this)
  }

  static async getInitialProps ({ reduxStore, req, res, query: { id } }) {
    const product = await ProductPage.fetchProduct(id, reduxStore.dispatch)
    const isServer = !!req
    const { sellable } = product
    const preview = req.query && req.query.preview ? req.query.preview : false

    // Unpublished products should return a 404 page to the public however we should
    // be able to preview unpublished products using a ?preview=true query string
    if (sellable === false && preview === false) {
      if (isServer) {
        // server-side
        res.redirect('/error')
      } else {
        // client-side
        Router.push('/error')
      }
    }

    return { id: id, product }
  }

  static async fetchProduct (id, dispatch) {
    try {
      const request = ProductPage.productRequest(id)
      const response = await new ApiClient().read(request.endpoint, request.query, dispatch)

      return response.data
    } catch (error) {
      return { error }
    }
  }

  static productRequest (productId) {
    return {
      endpoint: `/getProduct/${productId}`,
      query: {
        fields: {
          asset_files: 'id,image_height,image_width,s3_url',
          variants: 'id,meta_attributes,title,description,reference,ewis_eligible,product_id,sku,stock_allocated_level,stock_level,tax_code,stock_available_level,has_assets,price_includes_taxes,available_to_order,price,current_price,picture_url'
        },
        include: 'asset_files,variants,template,meta.*'
      }
    }
  }

  componentDidMount () {
    const { dispatch, id } = this.props
    dispatch(readProduct(id))
  }

  addToBag () {
    const { variantId, quantity } = this.state
    this.props.dispatch(addToCart(variantId, parseInt(quantity))).then(success => {
      if (success) this.props.dispatch(toggleMiniBag(true))
    })
  }

  changeVariant (e) {
    this.setState({
      variantId: e.target.attributes['data-variant-id'].value
    })
  }

  renderPageHead (title) {
    return (
      <this.Head>
        <title>{ suffixWithStoreName(title) }</title>
      </this.Head>
    )
  }

  /**
   * Render the product page when loaded. This method is seperate to the main
   * render method so it can be overridden, without overriding the loading and
   * error parts of the render method
   * @return {String} - HTML markup for the component
   */
  renderLoaded () {
    const { product, product: { template } } = this.props

    const templateSection = template && template.sections && template.sections.slice(-1).pop()
    const components = templateSection && templateSection.components
    const selectedVariant = product.variants.find(variant => variant.id === this.state.variantId)

    return (
      <Fragment>
        <ProductDisplay
          product={product}
          changeVariant={this.changeVariant}
          addToBag={this.addToBag}
          selectedVariant={selectedVariant}
        />
        { components && renderComponents(components) }
      </Fragment>
    )
  }

  render () {
    const { product, product: { loading, error, title } } = this.props

    if (loading) {
      return (
        <Loading />
      )
    } else if (error || Object.keys(product).length === 0) {
      return (
        <h1>Unable to load product.</h1>
      )
    } else {
      return (
        <Fragment>
          { this.renderPageHead(title) }
          { this.renderLoaded() }
        </Fragment>
      )
    }
  }
}

export default ProductPage
