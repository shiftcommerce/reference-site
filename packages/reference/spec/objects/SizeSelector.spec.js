// Object
import SizeSelector from '../../objects/SizeSelector'

// Fixtures
import product from '../fixtures/product.fixture'

test('renders correctly', () => {
  // Arrange & Act
  const wrapper = mount(
    <SizeSelector name='line_item[item_id]' prompt='Select a Size' variants={product.variants} />
  )

  // Assert
  expect(wrapper).toMatchSnapshot()
})
