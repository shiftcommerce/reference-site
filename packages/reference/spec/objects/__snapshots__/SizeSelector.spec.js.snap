// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<SizeSelector
  name="line_item[item_id]"
  prompt="Select a Size"
  variants={
    Array [
      Object {
        "meta_data": Object {
          "eu": Object {
            "price": 10,
            "size": "Size 8",
          },
        },
        "sku": 1,
        "title": "Small",
      },
      Object {
        "meta_data": Object {
          "eu": Object {
            "price": 11,
            "size": "Size 10",
          },
        },
        "sku": 2,
        "title": "Medium",
      },
      Object {
        "meta_data": Object {
          "eu": Object {
            "price": 12,
            "size": "Size 12",
          },
        },
        "sku": 3,
        "title": "Large",
      },
    ]
  }
>
  <select
    name="line_item[item_id]"
  >
    <option
      aria-posinset="0"
      aria-setsize={3}
      role="option"
      value=""
    >
      Select a Size
    </option>
    <option
      aria-posinset={1}
      aria-setsize={3}
      data-price={10}
      key="0"
      role="option"
      value={1}
    >
      Size 8
    </option>
    <option
      aria-posinset={2}
      aria-setsize={3}
      data-price={11}
      key="1"
      role="option"
      value={2}
    >
      Size 10
    </option>
    <option
      aria-posinset={3}
      aria-setsize={3}
      data-price={12}
      key="2"
      role="option"
      value={3}
    >
      Size 12
    </option>
  </select>
</SizeSelector>
`;
